### PR TITLE
feat(size): create new effect

### PR DESCRIPTION
**Pull request by johnleider**
## Description
[**Notion Task**](https://www.notion.so/vuetify/E-Sizeable-c1493a15953d4ea98d5d2ee0e379256e)

A compositional implementation of the size mixin.

## How Has This Been Tested?
Cannot unit test v3 alpha atm

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Improvement/refactoring (non-breaking change that doesn't add any features but makes things better)

## Checklist:

- [x] The PR title is no longer than 64 characters.
- [x] The PR is submitted to the correct branch (`master` for bug fixes and documentation updates, `dev` for new features and breaking changes).
- [x] My code follows the code style of this project.
- [ ] I've added relevant changes to the documentation (applies to new features and breaking changes in core library)

### CODE REVIEW

**Comment by johnleider** [comment]
Requesting initial comments @vuetifyjs/core-contributors @vuetifyjs/core-team

**Comment by jacekkarczmarczyk** [comment]
As per conding guidelines:

image

we should consider changing it to one `size` prop with values `"x-small"`/`"small"`/`"default"`|undefined/'"large"`/`"x-large"`

**Comment by johnleider** [comment]
Requesting **Final** comments from @vuetifyjs/contributors @vuetifyjs/core-team

File: packages/vuetify/src/effects/__tests__/__snapshots__/size.spec.ts.snap [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+exports[`size.ts should have the correct class 8`] = `
+Object {
+  "default": undefined,
+  "size": [Function],
+}
+`;
+
+exports[`size.ts should have the correct class 9`] = `
+Object {
+  "v-size--undefined": true,
```
Comment (jacekkarczmarczyk) [inline packages/vuetify/src/effects/__tests__/__snapshots__/size.spec.ts.snap:41]: is this expected?

File: packages/vuetify/src/effects/__tests__/__snapshots__/size.spec.ts.snap [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+}
+`;
+
+exports[`size.ts should have the correct class 10`] = `
+Object {
+  "default": undefined,
+  "size": [Function],
```
Comment (jacekkarczmarczyk) [inline packages/vuetify/src/effects/__tests__/__snapshots__/size.spec.ts.snap:48]: snapshot doesn't seem to be a good way to test this

**Comment by johnleider** [comment]
What about pivoting from xs -> xl and doing this: https://material.io/design/layout/applying-density.html#components

File: packages/vuetify/src/composables/size.ts [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+export interface SizeProps {
+  size: Size
+}
+
+const allowedSizes = ['x-small', 'small', 'default', 'large', 'x-large'] as const
+
+type Size = typeof allowedSizes[number] | undefined
+
+// Props
+export function sizeProps (defaults: Partial<SizeProps> = { size: 'default' }) {
```
Comment (KaelWD) [inline packages/vuetify/src/composables/size.ts:14]: Are we going with `fooProps` or `makeFooProps`?

File: packages/vuetify/src/composables/size.ts [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    } as Prop<Size>,
+  }
+}
+
+// Effect
+export function useSizeClasses (props: SizeProps) {
+  const sizeClasses = computed(() => {
+    if (!props?.size) return {}
+
+    return { [`v-size--${props.size}`]: true }
```
Comment (KaelWD) [inline packages/vuetify/src/composables/size.ts:29]: Classes can just be a string; usage:
```js
const { sizeClasses } = useSizeClasses(props)
return h('div', {
  class: ['static-class', sizeClasses, moreClasses, { something: condition }],
})
```

File: packages/vuetify/src/composables/__tests__/size.spec.ts [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+describe('size', () => {
+  it('should warn if value is incorrect', () => {
+    const TestComponent = {
+      props: sizeProps(),
+      template: '<div/>',
+    }
+
+    mount(TestComponent, {
+      props: {
+        // @ts-ignore
```
Comment (KaelWD) [inline packages/vuetify/src/composables/__tests__/size.spec.ts:13]:
```suggestion
        // @ts-expect-error
```

File: packages/vuetify/package.json [hunk 6]
```diff
@@ -70,8 +70,9 @@
     "@mdi/font": "^4.5.95",
     "@types/jest": "^24.9.1",
     "@types/node": "^12.6.8",
-    "@vue/test-utils": "^1.0.0-beta.29",
+    "@vue/test-utils": "^2.0.0-beta.7",
     "autoprefixer": "^9.6.1",
+    "babel-jest": "^26.3.0",
```
Comment (KaelWD) [inline packages/vuetify/package.json:75]: This is already in the root package.json

File: packages/vuetify/src/composables/size.ts (hunk 3 above)
Comment (nekosaur) [reply packages/vuetify/src/composables/size.ts:14]: make

File: packages/vuetify/src/composables/__tests__/size.spec.ts [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+    })
+
+    expect('[Vue warn]: Invalid prop: custom validator check failed for prop "size".').toHaveBeenTipped()
+  })
+
+  it.each([
+    [undefined, null],
+    [null, null],
+    [0 as any, null],
+    ['' as any, null],
```
Comment (KaelWD) [inline packages/vuetify/src/composables/__tests__/size.spec.ts:25]: `props` is never going to be any of these

File: packages/vuetify/src/composables/size.ts [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+      default: defaults.size,
+      validator: (v: any) => allowedSizes.includes(v),
+    } as Prop<Size>,
+  }
+}
+
+// Effect
+export function useSizeClasses (props: SizeProps) {
+  const sizeClasses = computed(() => {
+    if (!props?.size) return null
```
Comment (KaelWD) [inline packages/vuetify/src/composables/size.ts:29]: `props` is always at least `{}`

**Pull request merged**